ledger/complete/wal/checkpoint_v6_reader.go

```diff
@@ -269,11 +369,64 @@ func readTopLevelTries(dir string, fileName string, subtrieNodes [][]*node.Node,
 		tries[i] = trie
 	}
 
-	// TODO: validate checksum
+	// read footer and discard, since we only care about checksum
+	_, err = io.ReadFull(reader, scratch[:encNodeCountSize+encTrieCountSize])
+	if err != nil {
+		return nil, fmt.Errorf("cannot read footer: %w", err)
+	}
+
+	actualSum := reader.Crc32()
+
+	// read the stored checksum, and compare with the actual sum
+	_, err = readCRC32Sum(reader)
+	if err != nil {
+		return nil, fmt.Errorf("could not read top level trie checksum: %w", err)
+	}
+
+	if actualSum != expectedSum {
+		return nil, fmt.Errorf("invalid checksum in top level trie, expected %v, actual %v",
+			expectedSum, actualSum)
+	}
 
 	return tries, nil
 }
 
+func readVersion(reader io.Reader) (uint16, error) {
```

Comment:
See the changes to `readVersion` [in this PR](https://github.com/onflow/flow-go/pull/3335/files#diff-2ff6507ee260a1e37156b4cae4cc259e7688bbd0007993b20fa193824073105aR429).

In that PR, I let `readVersion` to take the magic ID, so that it will validate and return the version. 

And exactly, I did reuse this function for different magic ID. 

Thinking about this again, I could have
```
func readHeader(reader io.Reader) (uint16, uint16, error)
func validateHeader(magicID uint16, version uint16) error
```

`validateHeader` will call readHeader and validate the magic ID and version.